### PR TITLE
GDScript: Allow using local constants as types

### LINKED ISSUE #80960: Can't use local constant as type
### Godot version

4.2 dev3

### System information

Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1060 (NVIDIA; 30.0.15.1403) - Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz (8 Threads)

### Issue description

```GDScript
func _ready() -> void:
	const Type = preload("res://Script.gd")
	var test: Type
```
image

Works if type is outside function.

### Steps to reproduce

See above.

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by dalexeev**
* Closes #80960.

In master you can only use member constants as types. Now it's possible:

```gdscript
enum MyEnum {A, B, C}

func _ready() -> void:
    const X = preload("1.gd")
    const Y = MyEnum
    var x: X
    var y: Y
```

I recommend hide whitespace changes for the review, since I increased the indentation level in a large block.

### CODE REVIEW

**Comment by adamscott** [comment]
@dalexeev Is it cherry-pickable?

**Comment by dalexeev** [comment]
@adamscott I think yes, with a small change in the parser from #79880 (add the `IdentifierNode::suite` property). However, I think it's more of an inconvenience than a significant limitation (it's easy to workaround by using member constant instead of local constant). In addition, we regularly publish 4.2 dev snapshots. But if production team thinks it's important enough for a cherry-pick, then I don't mind.

File: modules/gdscript/gdscript_analyzer.cpp [hunk 1]
```diff
@@ -613,144 +613,169 @@ GDScriptParser::DataType GDScriptAnalyzer::resolve_datatype(GDScriptParser::Type
...
-		// Built-in types.
-		if (p_type->type_chain.size() > 1) {
-			push_error(R"(Built-in types don't contain nested types.)", p_type->type_chain[1]);
+			if (!result.is_meta_type) {
+				push_error(vformat(R"(Local constant "%s" is not a valid type.)", first), first_id);
+				return bad_type;
+			}
+			type_found = true;
+		} else {
+			push_error(vformat(R"(Identifier "%s" is a local %s, so it cannot be a valid type.)", first, local.get_name()), first_id);
```
Comment (dalexeev) [inline modules/gdscript/gdscript_analyzer.cpp:635]:
```suggestion
			push_error(vformat(R"(Local %s "%s" cannot be used as a type.)", local.get_name(), first), first_id);
```

**Review by adamscott — APPROVED** [review]
This seems nice! Let's test it out.

**Comment by YuriSizov** [comment]
Thanks!

**Pull request merged**